File: ReactNative/IMSDK/react-native-tim-js/src/manager/v2_tim_signaling_manager.ts
/**
 * 信令相关功能，主要用于音视频通话信令的发送。
 * @module SignalingManager(高级消息收发接口)
 */
import type V2TimCallback from '../interface/v2TimCallback';
import type V2TimSignalingInfo from 'src/interface/v2TimSignalingInfo';
import type V2TimOfflinePushInfo from '../interface/v2TimOfflinePushInfo';
import type V2TimValueCallback from '../interface/v2TimValueCallback';
import type V2TimSignalingListener from '../interface/v2TimSignalingListener';
import type { NativeEventEmitter } from 'react-native';

export class V2TIMSignalingManager {
    private manager: String = 'signalingManager';
    private nativeModule: any;
    private listenerList: V2TimSignalingListener[] = [];

    /** @hidden */
    constructor(module: any, eventEmitter: NativeEventEmitter) {
        this.nativeModule = module;
        eventEmitter.addListener('signalingListener', (response: any) => {
            const { type, data } = response;
            this.listenerList.forEach((listener) => {
                switch (type) {
                    case 'onReceiveNewInvitation':
                        listener.onReceiveNewInvitation &&
                            listener.onReceiveNewInvitation(
                                data.inviteID,
                                data.inviter,
                                data.groupID,
                                data.inviteeList,
                                data.data
                            );
                        break;
                    case 'onInviteeAccepted':
                        listener.onInviteeAccepted &&
                            listener.onInviteeAccepted(
                                data.inviteID,
                                data.invitee,
                                data.data
                            );
                        break;
                    case 'onInviteeRejected':
                        listener.onInviteeRejected &&
                            listener.onInviteeRejected(
                                data.inviteID,
                                data.invitee,
                                data.data
                            );
                        break;
                    case 'onInvitationCancelled':
                        listener.onInvitationCancelled &&
                            listener.onInvitationCancelled(
                                data.inviteID,
                                data.inviter,
                                data.data
                            );
                        break;
                    case 'onInvitationTimeout':
                        listener.onInvitationTimeout &&
                            listener.onInvitationTimeout(
                                data.inviteID,
                                data.inviteeList
                            );
                        break;
                }
            });
        });
    }

    /**
     * ### 添加信令监听
     */
    public addSignalingListener(listener: V2TimSignalingListener): void {
        if (this.listenerList.length == 0) {
            this.nativeModule.call(this.manager, 'addSignalingListener', {});
        }
        this.listenerList.push(listener);
    }
    /**
     * ### 移除信令监听
     */
    public removeSignalingListener(listener?: V2TimSignalingListener): void {
        if (!listener) {
            this.listenerList = [];
        } else {
            this.listenerList = this.listenerList.filter(
                (item) => item != listener
            );
        }
        if (this.listenerList.length == 0) {
            this.nativeModule.call(this.manager, 'removeSignalingListener', {});
        }
    }

    /**
     * 邀请某个人
     * @param invitee - 被邀请人用户 ID
     * @param data - 自定义数据
     * @param timeout - 超时时间，单位秒，如果设置为 0，SDK 不会做超时检测，也不会触发 onInvitationTimeout 回调
     * @param onlineUserOnly - 是否只有在线用户才能收到邀请，如果设置为 true，只有在线用户才能收到， 并且 invite 操作也不会产生历史消息（针对该次 invite 的后续 cancel、accept、reject、timeout 操作也同样不会产生历史消息）。
     * @param offlinePushInfo - 离线推送信息，其中 desc 为必填字段，推送的时候会默认展示 desc 信息。
     */
    public invite(
        invitee: String,
        data: String,
        timeout = 30,
        onlineUserOnly = false,
        offlinePushInfo?: V2TimOfflinePushInfo
    ): Promise<V2TimValueCallback<String>> {
        return this.nativeModule.call(this.manager, 'invite', {
            invitee,
            data,
            timeout,
            onlineUserOnly,
            offlinePushInfo,
        });
    }

    /**
     * ### 邀请群内的某些人
     * @param groupID - 发起邀请所在群组
     * @param inviteeList - 被邀请人列表，inviteeList 必须已经在 groupID 群里，否则邀请无效
     * @param data - 自定义数据
     * @param timeout - 超时时间，单位秒，如果设置为 0，SDK 不会做超时检测，也不会触发 onInvitationTimeout 回调
     * @param onlineUserOnly - 是否只有在线用户才能收到邀请，如果设置为 true，只有在线用户才能收到， 并且 invite 操作也不会产生历史消息（针对该次 invite 的后续 cancel、accept、reject、timeout 操作也同样不会产生历史消息）。
     * @note
     * 群邀请暂不支持离线推送，如果您需要离线推送，可以针对被邀请的用户单独发离线推送自定义消息，
     */
    public inviteInGroup(
        groupID: String,
        inviteeList: String[],
        data: String,
        timeout = 30,
        onlineUserOnly = false
    ): Promise<V2TimValueCallback<String>> {
        return this.nativeModule.call(this.manager, 'inviteInGroup', {
            groupID,
            inviteeList,
            data,
            timeout,
            onlineUserOnly,
        });
    }

    /**
     * ### 邀请方取消邀请
     * @note
     * 如果所有被邀请人都已经处理了当前邀请（包含超时），不能再取消当前邀请。
     */
    public cancel(inviteID: String, data?: String): Promise<V2TimCallback> {
        return this.nativeModule.call(this.manager, 'cancel', {
            inviteID,
            data,
        });
    }

    /**
     * ### 接收方接收邀请
     * @note
     * 不能接受不是针对自己的邀请，请在收到 onReceiveNewInvitation 回调的时候先判断 inviteeList 有没有自己，如果没有自己，不能 accept 邀请。
     */
    public accept(inviteID: String, data?: String): Promise<V2TimCallback> {
        return this.nativeModule.call(this.manager, 'accept', {
            inviteID,
            data,
        });
    }

    /**
     * ### 接收方拒绝邀请
     * @note
     * 不能拒绝不是针对自己的邀请，请在收到 onReceiveNewInvitation 回调的时候先判断 inviteeList 有没有自己，如果没有自己，不能 reject 邀请。
     */
    public reject(inviteID: String, data?: String): Promise<V2TimCallback> {
        return this.nativeModule.call(this.manager, 'reject', {
            inviteID,
            data,
        });
    }

    /**
     * ### 获取信令信息
     * 如果 invite 设置 onlineUserOnly 为 false，每次信令操作（包括 invite、cancel、accept、reject、timeout）都会产生一条自定义消息， 该消息会通过 V2TIMAdvancedMsgListener -> onRecvNewMessage 抛给用户，用户也可以通过历史消息拉取，如果需要根据信令信息做自定义化文本展示，可以调用下面接口获取信令信息。
     */
    public getSignalingInfo(
        msgID: String
    ): Promise<V2TimValueCallback<V2TimSignalingInfo>> {
        return this.nativeModule.call(this.manager, 'getSignalingInfo', {
            msgID,
        });
    }

    /**
     * ### 添加邀请信令（可以用于群离线推送消息触发的邀请信令）
     * 在离线推送的场景下：
     * - 针对 1V1 信令邀请，被邀请者 APP 如果被 Kill，当 APP 收到信令离线推送再次启动后，SDK 可以自动同步到邀请信令，如果邀请还没超时， 用户会收到 onReceiveNewInvitation 回调，如果邀请已经超时，用户还会收到 onInvitationTimeout 回调。
     * - 针对群信令邀请，被邀请者 APP 如果被 Kill，当 APP 收到离线推送再次启动后，SDK 无法自动同步到邀请信令（邀请信令本质上就是一条自定义消息，群离线消息在程序启动后无法自动同步）， 也就没法处理该邀请信令。如果被邀请者需要处理该邀请信令，可以让邀请者在发起信令的时候对针对每个被邀请者再单独发送一条 C2C 离线推送消息，消息里面携带 V2TIMSignalingInfo 信息， 被邀请者收到离线推送的时候把 V2TIMSignalingInfo 信息再通过 addInvitedSignaling 接口告知 SDK。
     * - TUIKit 音视频通话离线推送功能基于这个接口实现，详细实现方法请参考文档：[集成音视频通话](https://cloud.tencent.com/document/product/647)
     */
    public addInvitedSignaling(
        info: V2TimSignalingInfo
    ): Promise<V2TimCallback> {
        return this.nativeModule.call(this.manager, 'addInvitedSignaling', {
            info,
        });
    }
}

File: Flutter/IMSDK/im-flutter-plugin/tencent_im_sdk_plugin_web/lib/src/manager/v2_tim_signaling_manager.dart
import 'dart:async';
import 'dart:convert';

import 'package:tencent_im_sdk_plugin_platform_interface/enum/V2TimSignalingListener.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/enum/offlinePushInfo.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/enum/v2_signaling_action_type.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_callback.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_message.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_msg_create_info_result.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_signaling_info.dart';
import 'package:tencent_im_sdk_plugin_platform_interface/models/v2_tim_value_callback.dart';
import 'package:tencent_im_sdk_plugin_web/tencent_im_sdk_plugin_web.dart';
import 'package:uuid/uuid.dart';

class V2TIMSignalingManager {
  ///@nodoc
  Map<String, V2TimSignalingListener> signalingListenerList = {};
  Map<String, Timer> timeoutTimer = {};
  Map<String, dynamic> currentInvite = {};

  ///添加信令监听
  ///
  Future<void> addSignalingListenerForWeb({
    required V2TimSignalingListener listener,
    required String listenerUuid,
  }) {
    signalingListenerList[listenerUuid] = listener;
    return Future.value();
  }

  ///移除信令监听
  ///
  Future<void> removeSignalingListener({
    V2TimSignalingListener? listener,
    required String listenerUuid,
  }) {
    var listenerUuid = "";
    if (listener != null) {
      listenerUuid = signalingListenerList.keys.firstWhere(
          (k) => signalingListenerList[k] == listener,
          orElse: () => "");
      signalingListenerList.remove(listenerUuid);
    } else {
      signalingListenerList.clear();
    }
    return Future.value();
  }

  Future<int> _sendCustomData(
      {required String data,
      String? groupID,
      String? userID,
      OfflinePushInfo? pusinfo,
      bool? onlineUserOnly}) async {
    V2TimValueCallback<V2TimMsgCreateInfoResult> res =
        await TencentImSDKPluginWeb().createCustomMessage(data: data,extension: "",desc: "");
    if (res.data != null) {
      String id = res.data!.id!;
      V2TimValueCallback<V2TimMessage> callback = await TencentImSDKPluginWeb()
          .sendMessage(
              id: id,
              receiver: userID ?? "",
              groupID: groupID ?? "",
              offlinePushInfo: pusinfo?.toJson() ?? OfflinePushInfo().toJson(),
              onlineUserOnly: onlineUserOnly ?? false);
      return callback.code;
    }
    return -1;
  }

  _timeout(int timeout, String inviteID) async {
    if (timeout > 0) {
      timeoutTimer[inviteID] =
          Timer.periodic(Duration(seconds: timeout), (timer) async {
        if (timeoutTimer[inviteID] != null) {
          if (timeoutTimer[inviteID]!.isActive) {
            timeoutTimer[inviteID]!.cancel();
            // check if there is a avilabal signal
            if (currentInvite["inviteID"] != null) {
              V2TimValueCallback<String> user =
                  await TencentImSDKPluginWeb().getLoginUser();
              // send a timeout message
              List<String> inviteeList =
                  List<String>.from(currentInvite["inviteeList"]);
              if (inviteeList.isNotEmpty) {
                String data = _buildSignalData(
                  type: V2SignalingActionType
                      .SIGNALING_ACTION_TYPE_INVITE_TIMEOUT,
                  customData: currentInvite["data"],
                  inviteID: currentInvite["inviteID"],
                  timeout: currentInvite["timeout"],
                  groupID: currentInvite["groupID"],
                  inviter: currentInvite["inviter"],
                  inviteeList: inviteeList,
                );
                print("send timeout message");
                await _sendCustomData(
                  data: data,
                  groupID: currentInvite["groupID"] ?? "",
                  userID: _isGroupCall(currentInvite) ? "" : inviteeList.first,
                );
                signalingListenerList.forEach((key, value) {
                  value.onInvitationTimeout(inviteID,inviteeList);
                });
                currentInvite = {};
              }
            }
          }
        }
      });
    }
  }

  bool _isGroupCall(Map<String, dynamic> callinfo) {
    if (callinfo["groupID"] == null) {
      return false;
    } else {
      if (callinfo["groupID"].isNotEmpty) {
        return true;
      }
      return false;
    }
  }

  _onCallbackTimeout(Map<String, dynamic> callinfo) async {
    int timeout = callinfo["timeout"] ?? 30;
    String inviteID = callinfo["inviteID"] ?? "";
    if (timeout >= 0 && inviteID.isNotEmpty) {
      V2TimValueCallback<String> user =
          await TencentImSDKPluginWeb().getLoginUser();
      // The countdown on the receiver side is two seconds less because of network transfer time
      timeoutTimer[inviteID] = Timer.periodic(
          Duration(seconds: timeout - 2 > 0 ? timeout - 2 : timeout), (timer) {
        // List<String> inviteeList = callinfo["inviteeList"];
        callinfo["inviteeList"] = [user.data];
        _sendCustomData(
            data: json.encode(callinfo),
            userID: _isGroupCall(callinfo) ? "" : callinfo["inviter"],
            groupID: callinfo["groupID"]);
      });
    }
  }

  beforeCallback(int type, Map<String, dynamic> callinfo) async {
    switch (type) {
      case V2SignalingActionType.SIGNALING_ACTION_TYPE_ACCEPT_INVITE:
        List<String> inviteeList = List<String>.from(callinfo["inviteeList"]);
        if (inviteeList.isNotEmpty) {
          List<String> currentInviteeList =
              List<String>.from(currentInvite["inviteeList"]);
          for (var element in inviteeList) {
            currentInviteeList.remove(element);
          }
          currentInvite["inviteeList"] = currentInviteeList;
          if (currentInviteeList.isEmpty) {
            currentInvite = {};
          }
        }
        break;
      case V2SignalingActionType.SIGNALING_ACTION_TYPE_CANCEL_INVITE:
        String inviteID = callinfo["callinfo"] ?? "";
        if (inviteID.isNotEmpty) {
          currentInvite = {};
        }
        break;
      case V2SignalingActionType.SIGNALING_ACTION_TYPE_INVITE:
        // handle on invited and timeout
        currentInvite = callinfo;
        _onCallbackTimeout(callinfo);
        break;
      case V2SignalingActionType.SIGNALING_ACTION_TYPE_INVITE_TIMEOUT:
        List<String> inviteeList = List<String>.from(callinfo["inviteeList"]);
        if (inviteeList.isNotEmpty) {
          List<String> currentInviteeList =
              List<String>.from(currentInvite["inviteeList"]);
          for (var element in inviteeList) {
            currentInviteeList.remove(element);
          }
          currentInvite["inviteeList"] = currentInviteeList;
          if (currentInviteeList.isEmpty) {
            currentInvite = {};
          }
        }
        break;
      case V2SignalingActionType.SIGNALING_ACTION_TYPE_REJECT_INVITE:
        List<String> inviteeList = List<String>.from(callinfo["inviteeList"]);
        if (inviteeList.isNotEmpty) {
          List<String> currentInviteeList =
              List<String>.from(currentInvite["inviteeList"]);
          for (var element in inviteeList) {
            currentInviteeList.remove(element);
          }
          currentInvite["inviteeList"] = currentInviteeList;
          if (currentInviteeList.isEmpty) {
            currentInvite = {};
          }
        }
        break;
    }
  }

  String _buildSignalData(
      {required int type,
      required String customData,
      required String inviteID,
      required int timeout,
      String? groupID,
      required List<String> inviteeList,
      required String inviter}) {
    Map<String, dynamic> signal = {
      "businessID": 1,
      "inviteID": inviteID,
      "data": customData,
      "inviter": inviter,
      "actionType": type,
      "inviteeList": inviteeList,
      "timeout": timeout,
      "groupID": groupID ?? "",
    };
    return json.encode(signal);
  }

  Future<V2TimValueCallback<String>> invite({
    required String invitee,
    required String data,
    int timeout = 30,
    bool onlineUserOnly = false,
    OfflinePushInfo? offlinePushInfo,
  }) async {
    // send custom message and handle timeout.
    V2TimValueCallback<String> user =
        await TencentImSDKPluginWeb().getLoginUser();
    String inviteID = const Uuid().v4();
    String customData = _buildSignalData(
      type: V2SignalingActionType.SIGNALING_ACTION_TYPE_INVITE,
      customData: data,
      inviteID: inviteID,
      timeout: timeout,
      groupID: "",
      inviteeList: [invitee],
      inviter: user.data ?? "",
    );
    int code = await _sendCustomData(
      data: customData,
      userID: invitee,
      pusinfo: offlinePushInfo,
      onlineUserOnly: onlineUserOnly,
    );
    currentInvite = json.decode(customData);
    if (code == 0) {
      _timeout(timeout, inviteID);
    }
    return V2TimValueCallback(
        code: code, data: inviteID, desc: code == 0 ? "success" : "error");
  }

  Future<V2TimValueCallback<String>> inviteInGroup({
    required String groupID,
    required List<String> inviteeList,
    required String data,
    int timeout = 30,
    bool onlineUserOnly = false,
  }) async {
    // send custom message and handle timeout.
    String inviteID = const Uuid().v4();
    V2TimValueCallback<String> user =
        await TencentImSDKPluginWeb().getLoginUser();
    String customData = _buildSignalData(
        type: V2SignalingActionType.SIGNALING_ACTION_TYPE_INVITE,
        customData: data,
        inviteID: inviteID,
        timeout: timeout,
        groupID: groupID,
        inviteeList: inviteeList,
        inviter: user.data ?? "");
    int code = await _sendCustomData(
      data: customData,
      groupID: groupID,
      onlineUserOnly: onlineUserOnly,
    );
    currentInvite = json.decode(customData);
    if (code == 0) {
      _timeout(timeout, inviteID);
    }
    return V2TimValueCallback(
        code: code, data: inviteID, desc: code == 0 ? "success" : "error");
  }

  Future<V2TimCallback> cancel({
    required String inviteID,
    String? data,
  }) async {
    if (currentInvite[inviteID] != null) {
      List<String> inviteeList =
          List<String>.from(currentInvite["inviteeList"]);
      String customData = _buildSignalData(
          type: V2SignalingActionType.SIGNALING_ACTION_TYPE_CANCEL_INVITE,
          customData: currentInvite["data"],
          inviteID: inviteID,
          timeout: currentInvite["timeout"],
          groupID: currentInvite["groupID"],
          inviteeList: inviteeList,
          inviter: currentInvite["inviter"]);
      if (inviteeList.isNotEmpty) {
        int code = await _sendCustomData(
          data: customData,
          groupID: currentInvite["groupID"] ?? "",
          userID: _isGroupCall(currentInvite) ? "" : inviteeList.first,
        );
        currentInvite[inviteID] = null;
        if (timeoutTimer[inviteID] != null) {
          if (timeoutTimer[inviteID]!.isActive) {
            timeoutTimer[inviteID]!.cancel();
          }
        }
        return V2TimCallback(code: code, desc: code == 0 ? "success" : "error");
      }
    }
    return V2TimCallback(
          code: 8010,
          desc: "inviteID is invalid or invitation has been processed");
  }

  Future<V2TimCallback> accept({
    required String inviteID,
    String? data,
  }) async {
    if (currentInvite[inviteID] != null) {
      V2TimValueCallback<String> user =
          await TencentImSDKPluginWeb().getLoginUser();
      String customData = _buildSignalData(
          type: V2SignalingActionType.SIGNALING_ACTION_TYPE_ACCEPT_INVITE,
          customData: currentInvite["data"],
          inviteID: inviteID,
          timeout: currentInvite["timeout"],
          groupID: currentInvite["groupID"],
          inviteeList: [user.data!],
          inviter: currentInvite["inviter"],
          );
      int code = await _sendCustomData(
        data: customData,
        groupID: currentInvite["groupID"] ?? "",
        userID: _isGroupCall(currentInvite) ? "" : currentInvite["inviter"],
      );
      currentInvite[inviteID] = null;
      if (timeoutTimer[inviteID] != null) {
        if (timeoutTimer[inviteID]!.isActive) {
          timeoutTimer[inviteID]!.cancel();
        }
      }

      return V2TimCallback(code: code, desc: code == 0? "success":"error");
    } 
    return V2TimCallback(
          code: 8010,
          desc: "inviteID is invalid or invitation has been processed");
  }

  Future<V2TimCallback> reject({
    required String inviteID,
    String? data,
  }) async {
    if (currentInvite[inviteID] != null) {
      V2TimValueCallback<String> user =
          await TencentImSDKPluginWeb().getLoginUser();
      String customData = _buildSignalData(
          type: V2SignalingActionType.SIGNALING_ACTION_TYPE_REJECT_INVITE,
          customData: currentInvite["data"],
          inviteID: inviteID,
          timeout: currentInvite["timeout"],
          groupID: currentInvite["groupID"],
          inviteeList: [user.data!],
          inviter: currentInvite["inviter"]);
      int code = await _sendCustomData(
        data: customData,
        groupID: currentInvite["groupID"] ?? "",
        userID: _isGroupCall(currentInvite) ? "" : currentInvite["inviter"],
      );
      currentInvite[inviteID] = null;
      if (timeoutTimer[inviteID] != null) {
        if (timeoutTimer[inviteID]!.isActive) {
          timeoutTimer[inviteID]!.cancel();
        }
      }
      return V2TimCallback(code: code, desc: code ==0 ? "success":"error");
    } 
    return V2TimCallback(
          code: 8010,
          desc: "inviteID is invalid or invitation has been processed");
  }

  /// 获取信令信息
  ///
  /// 如果 invite 设置 onlineUserOnly 为 false，每次信令操作（包括 invite、cancel、accept、reject、timeout）都会产生一条自定义消息， 该消息会通过 V2TIMAdvancedMsgListener -> onRecvNewMessage 抛给用户，用户也可以通过历史消息拉取，如果需要根据信令信息做自定义化文本展示，可以调用下面接口获取信令信息。
  ///
  /// 参数
  /// msg	消息对象
  ///
  /// 返回
  /// V2TIMSignalingInfo 信令信息，如果为 null，则 msg 不是一条信令消息。
  ///
  Future<V2TimValueCallback<V2TimSignalingInfo>> getSignalingInfo({
    required String msgID,
  }) async {
    return V2TimValueCallback<V2TimSignalingInfo>.fromJson({});
  }

  Future<V2TimCallback> addInvitedSignaling({
    required V2TimSignalingInfo info,
  }) async {
    return V2TimCallback.fromJson({});
  }
}
